### PR TITLE
fix(cache): replace in-memory table with an lru-cache to limit memory consumption

### PR DESCRIPTION
**Pull request by Tieske**
As the intent is to replace the caching all together, I did not make its size configurable

See code comments on some rough calculations. Configurability can be added if required, feedback welcome

### CODE REVIEW

**Comment by p0pr0ck5** [comment]
How hard would it be to pull our our implementation of TTLs with lua-resty-lrucache's TTL mechanism?

File: kong/tools/database_cache.lua [hunk 1]
```diff
@@ -2,12 +2,19 @@ local utils = require "kong.tools.utils"
 local resty_lock = require "resty.lock"
 local json_encode = require("cjson.safe").encode
 local json_decode = require("cjson.safe").decode
+local lrucache = require "resty.lrucache"
```
Comment (p0pr0ck5) [inline kong/tools/database_cache.lua]: what do we expect the average cache hit ratio to look like? if we expect a lot of cache misses and purges, should we consider `resty.lrucache.pureffi` (or making it configurable?)

Comment (Tieske) [inline kong/tools/database_cache.lua]: Depends on customer use. I'd expect it to be relatively high, based on "sessions", what I mean is that you'll most likely get multiple requests for a specific use. In that case the apis, cunsomers, credentials etc. will be the same for that session. Only when done the next session comes up.

but as said very speculative.

**Comment by Tieske** [comment]
> How hard would it be to pull our our implementation of TTLs with lua-resty-lrucache's TTL mechanism?

Impossible without modifying the lru library. We need to share the ttl between the lru cache and the shm cache. But if we get an existing entry from shm, we cannot get its remaining ttl. Hence we must have our own ttl field inserted.

**Comment by thefosk** [comment]
The size should still be configurable. We currently support [`mem_cache_size`](https://github.com/Mashape/kong/blob/master/kong.conf.default#L79) in the configuration file, so it should be configurable from there (or another property)?

**Comment by thibaultcha** [comment]
How about using the same value as `mem_cache_size`? There should however, be an upper limit as we know the LuaJIT VM memory limit is quite low, but as long as we are in a `1...500mb` range, I think it's reasonably to have the shm and the Lua-land LRU cache to be the same size. Thoughts?

**Comment by p0pr0ck5** [comment]
The value of `mem_cache_size` doesn't quite make sense, because `lrucache` wants a number of items, not a size limit, so the limits of the shdict and the lrucache will not be the same. And given that the usage of the lrucache will vary based entity size, and isn't something that can easily be documented, I think making this configurable will only confuse users. I'd vote to keep it small (100MB max per worker) and leave it alone- it's a transparent L1 cache, and doesn't need to be exposed on such a granular level to users.

**Comment by thefosk** [comment]
Also, `mem_cache_size` was being used once in the master process, while the LRU cache has a size limit for each worker. Which means that the actual total memory used would be:

`(mem_cache_size * num_workers) + master_shd_size`

**Comment by thibaultcha** [comment]
The same way we compute the number of items in the current patch with a hard-coded estimated item size, we could just as well do that by replacing this PR's `MEM_SIZE` with `mem_cache_size`.

In any case, I'm even more in favor of simply hard-coding this value.

**Comment by Tieske** [comment]
> There should however, be an upper limit as we know the LuaJIT VM memory limit is quite low, but as long as we are in a 1...500mb range

That's exactly why I picked the current size. The item size of 1024 is probably a lot bigger than we actually use. The growth we suspect is mostly based on the db-cache misses (standard data does not grow, it, at its maximum, consists of the entire datastore contents, and never more)
The db-miss sentinels are small, so my guess is that the maximum memory consumption per worker will be less than 250mb anyway.

> In any case, I'm even more in favor of simply hard-coding this value.

so am I.

**Pull request merged**